ui/components/src/TimeSeriesTooltip/nearby-series.ts

```diff
@@ -36,10 +37,151 @@ export interface NearbySeriesInfo {
 export type NearbySeriesArray = NearbySeriesInfo[];
 
 /**
- * Returns formatted series data for the points that are close to the user's cursor
+ * Returns formatted series data for the points that are close to the user's cursor.
+ * Adjust xBuffer and yBuffer to increase or decrease number of series shown.
+ */
+export function checkforNearbyTimeSeries(
+  data: TimeSeries[],
+  seriesMapping: TimeChartSeriesMapping,
+  pointInGrid: number[],
+  yBuffer: number,
+  chart?: EChartsInstance,
+  unit?: UnitOptions
+): NearbySeriesArray {
+  const currentNearbySeriesData: NearbySeriesArray = [];
+  const cursorX: number | null = pointInGrid[0] ?? null;
+  const cursorY: number | null = pointInGrid[1] ?? null;
+
+  if (cursorX === null || cursorY === null) return currentNearbySeriesData;
+
+  if (!Array.isArray(data)) return currentNearbySeriesData;
+
+  // TODO: better way to calc xBuffer for longer time ranges
+  // const xBuffer = cursorX * 0.0000005;
+
+  const nearbySeriesIndexes: number[] = [];
+  const emphasizedSeriesIndexes: number[] = [];
+  const nonEmphasizedSeriesIndexes: number[] = [];
+  const totalSeries = data.length;
+
+  let closestTimestamp = null;
+  let closestDistance = Infinity;
+
+  // find the timestamp with data that is closest to cursorX
+  for (let seriesIdx = 0; seriesIdx < totalSeries; seriesIdx++) {
+    const currentSeries = seriesMapping[seriesIdx];
+    if (currentSeries === undefined) break;
+
+    const currentDataset = totalSeries > 0 ? data[seriesIdx] : null;
+    if (currentDataset == null) break;
+
+    const currentDatasetValues: TimeSeriesValueTuple[] = currentDataset.values;
+
+    for (const [timestamp] of currentDatasetValues) {
```

Comment:
It might be possible to move this logic inside the for loop below so that we're only looping through currentDatasetValues once instead of twice

Reply:
nvm I see now, we do need to determine closestTimestamp _before_ the second loop that compares with xValue